File: db/src/main/java/onl/ycode/stormify/PopulationContext.java
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis

package onl.ycode.stormify;

import java.util.HashMap;
import java.util.Map;

class PopulationContext {
    private final Map<String, AutoTable> dedupMap = new HashMap<>();
    private final Map<Class<?>, SiblingGroup> siblingGroups = new HashMap<>();

    AutoTable getOrCreateReference(Class<?> type, Object idValue, ClassRegistry registry) {
        String key = type.getName() + ":" + idValue;
        AutoTable existing = dedupMap.get(key);
        if (existing != null)
            return existing;
        try {
            AutoTable wrapper = (AutoTable) type.getDeclaredConstructor().newInstance();
            registry.getTableInfo(type).getPrimaryKey().setValue(wrapper, idValue);
            dedupMap.put(key, wrapper);
            SiblingGroup group = siblingGroups.computeIfAbsent(type, k -> new SiblingGroup());
            group.add(wrapper);
            wrapper.siblingGroup = group;
            return wrapper;
        } catch (Exception e) {
            throw new QueryException("Failed to create reference for " + type.getSimpleName() + " with id " + idValue, e);
        }
    }
}

File: stormify/src/commonMain/kotlin/onl/ycode/stormify/PopulationContext.kt
// SPDX-License-Identifier: Apache-2.0
// (C) Panayotis Katsaloulis
package onl.ycode.stormify

import kotlin.reflect.KClass

internal class PopulationContext {
    private val dedupMap = mutableMapOf<String, Any>()
    private val siblingGroups = mutableMapOf<KClass<*>, SiblingGroup>()

    fun getOrCreateReference(
        type: KClass<*>,
        idValue: Any,
        stormify: Stormify
    ): Any {
        val key = "${type.fullName}:$idValue"
        dedupMap[key]?.let { return it }

        val wrapper = stormify.createReferenceStub(type, idValue)
        dedupMap[key] = wrapper

        if (wrapper is AutoTable) {
            val group = siblingGroups.getOrPut(type) { SiblingGroup() }
            group.add(wrapper)
            wrapper._siblingGroup = group
        }

        return wrapper
    }
}
